mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxViewModelFactory.kt

```diff
@@ -4,26 +4,84 @@ import android.support.v4.app.Fragment
 import android.support.v4.app.FragmentActivity
 
 /**
- * Implement this in the companion object of a MvRxViewModel if your ViewModel needs more dependencies than just initial state.
- * If all you need is initial state, you don't need to implement this at all.
+ * Implement this on your ViewModel's companion object for hooks into state creation and ViewModel creation. For example, if you need access
+ * to the fragment or activity owner for dependency injection.
  */
-interface MvRxViewModelFactory<S : MvRxState> {
+interface MvRxViewModelFactory<VM: BaseMvRxViewModel<S>, S : MvRxState> {
+
     /**
-     * This will be called when your ViewModel needs to created. This *needs* to be annotated with [JvmStatic].
-     * @param state: The initial state for your ViewModel. This will be populated from activity persisted state.
+     * @param viewModelContext [ViewModelContext] which contains the ViewModel owner and arguments.
+     * @param state The initial state to pass to the ViewModel. In a new process, state will have all [PersistState] annotated members restored,
+     * therefore you should never create a custom state in this method. To customize the initial state, override [initialState].
+     *
+     * @return The ViewModel. If you return `null` the ViewModel must have a single argument
+     * constructor only taking the initial state.
      */
-    fun create(activity: FragmentActivity, state: S): BaseMvRxViewModel<S>
+    fun create(viewModelContext: ViewModelContext, state: S): VM? = null
```

Comment:
The issue I ran into with having a single create in #148 was casting. If the consumer wants the `FragmentViewModelContext` then you need to downcast the `ViewModelContext`. 

For me, cause Dagger injects the `ViewModel` Factory implementation into the fragment, I would need to downcast for virtually all my `ViewModel`s factories to use the `FragmentViewModelContext`.

Reply:
@marukami Yeah, that is a downside, but I don't see a safe way around this. It seems too risky to have a ViewModelFactory that will crash if you use an activity scope.

I think, if in your application you know that you will only use fragment scoped view models, you could do something like:

```
interface MvRxViewModelFragmentFactory<VM: BaseMvRxViewModel<S>, S : MvRxState> : MvRxViewModelFactory<VM, S> {
    
    fun createFragmentScopedViewModel(viewModelContext: FragmentViewModelContext, state: S) : VM?
    
    override fun create(viewModelContext: ViewModelContext, state: S) = createFragmentScopedViewModel(viewModelContext as FragmentViewModelContext, state)

    override fun initialState(viewModelContext: ViewModelContext) : S? = initialStateFragmentScoped(viewModelContext as FragmentViewModelContext)

    fun initialStateFragmentScoped(viewModelContext: FragmentViewModelContext) : S? 
}
```

That way you are explicitly performing the unsafe cast. Does this seem reasonable?

Reply:
With the changes @gpeal suggestion for the Fragment been a generic, I think an extension function like this should work nicely. 

```kotlin
@Suppress("UNCHECKED_CAST")
inline fun <F: Fragment> ViewModelContext.withFragmentFactory(factory: (F) -> ViewModel) =
        (this as FragmentViewModelContext).let { it.fragment as F }.let(factory)
```